Added info on dependency version constraints (#5431)
dart-lang/site-www: src/tools/pub/dependencies.md

## Best practices

Be proactive in managing your dependencies.
Ensure that your packages use the freshest versions possible.

> Any stale dependencies could themselves depend on other stale packages.

Reviewer: This is a nice simplification of this text, but I'm worried it loses its meaning. It seems the original is actually talking about the other direction in which various packages in your dependency graph depend on that (stale) package. But if you are depending on a stale version of that package, the version solver might end up having to pull stale versions of your other dependencies, since their newer versions aren't compatible with the stale version you're depending on.

It's a bit intricate and this new direction is also a slight concern, so maybe it's fine to lose that meaning?

Reviewer 2: ```suggestion
A stale dependency could mean stale versions of other packages in your dependency graph
that depend on that package. 
```
To add back the correct relationship (I think?)

aren't on the latest stable versions,
use [`dart pub outdated`][].

> ### Use tight dev_dependencies

Reviewer: We tend to use "dependency" to refer to the package depended on, so it's not really the dependencies themselves that are "loose" or "tight", but rather the version constraint we specify for each of them.

Perhaps: 

```suggestion
### Use tight version constraints for dev dependencies
```

Reviewer 2: ```suggestion
### Tighten version constraints for dev dependencies
```

Dependencies can be loose, specifying dependency at the highest version digit
like `^2.0.0`, or tight, specifying dependency at the lowest version digit like

> `^1.2.4`.

Reviewer: I don't think we should necessarily define this in terms of which "version digit" as that's not a term we've used on the site. If someone is reading the best practices I think we can assume prerequisite knowledge of what each portion of the version specifier means, so let's specify it in terms of that (major, minor, patch).

Also, as mentioned above, it's best to refer to the constraints being tight or loose.

So in the case of a tight version constraint, that means the lower bound being the newest compatible version.

Reviewer 2: I'd skip the "loose" definition here if the best practice is just "tighten", define a "tight" constraint ("_that means the lower bound being the newest compatible version._"), and like Parker said [below](https://github.com/dart-lang/site-www/pull/5431/files#r1449864165), add the reasoning and definition of the reasoning to the opening description of the best practice, e.g.

    Specify dev dependencies at their lower bound, or newest compatible version,
    to ... (what does this achieve / avoid?)  
    OR
    because ... (something something transitive concerns?)

> In your `pubspec.yaml` file, follow these recommendations.

Reviewer: I'm not sure we need to mention this as this whole section is about dependencies in the `pubspec.yaml` file and recommendations for those dependencies. I think we can just focus on the best practices.

> * Set regular `dependencies` loose to increase the chance of a version solver success.

Reviewer: This section's title only mentions dev dependencies. Rather than expanding the focus of this section, I'd consider dropping this as the best practices for normal dependencies are much more nuanced and very situation dependent.

That's not to say we shouldn't have guidelines for them, but if we do have guidelines, we have to make them very clear.

> * Set [`dev_dependencies`][dev-dep] tighter as they lack transitive concerns.

Reviewer: If we can drop the note on normal dependencies, I think we should just combine this with the above text to better colocate what a tight dependency constraint is and why you should use them for dev dependencies.

Reviewer: Also if we keep using the wording "transitive concerns", could we add an example or two of those concerns? I'm afraid the terminology will be less familiar and potentially unclear about what concerns are being referred to.

I'd be happy to help expand here if you need any ideas. Let me know!

Author: It's at the top of the page.

> ```yaml

Reviewer: This code block confused me when reading as I didn't know why it was there. I think it can be dropped, but if you want to keep it, do add some more introduction/content, and perhaps some inline comments.

I also recommend using a non-Flutter example that doesn't use `sdk` dependencies, as they are more obscure and perhaps less familiar or completely unknown to non-Flutter developers.

aren't on the latest stable versions,
use [`dart pub outdated`][].

### Tighten version constraints for dev_dependencies

A dev dependency retrieves packages that you need only when developing.
These packages aren't needed in a finished app, like tests or examples.
Set the versions of package in [`dev_dependencies`][dev-dep]
to the exact version on which your package depends.

An example of how to set versions for dependencies would resemble the

> following `pubspec.yaml` example.

Author: Fixed.
